core/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -1,166 +1,108 @@
+    self orElse that
 }
 
 object ConfigSource {
   val SystemEnvironment = "system environment"
   val SystemProperties  = "system properties"
-  val JavaProperties    = "java properties"
-  val ConstantMap       = "constant <map>"
-  val EmptySource       = "<empty>"
 
   def empty[K, V]: ConfigSource[K, V] =
-    ConfigSource((k: Vector[K]) => IO.fail(singleton(ReadError.MissingValue(k))), EmptySource :: Nil)
+    ConfigSource(_ => PropertyTree.empty, Nil)
 
-  def fromEnv(valueSeparator: Option[String] = None): ConfigSource[String, String] =
-    ConfigSource(
-      (path: Vector[String]) => {
-        val key = path.mkString("_")
-        system
-          .env(key)
-          .bimap(
-            ReadError.FatalError(path, _),
-            opt =>
-              opt.map(r => {
-                val consOfValues: ::[String] =
-                  valueSeparator
-                    .flatMap(
-                      sep =>
-                        r.split(sep).toList.map(_.trim) match {
-                          case h :: tail => Some(::(h, tail))
-                          case Nil       => None
-                        }
-                    )
-                    .getOrElse(::(r, Nil))
-
-                ConfigValue(consOfValues)
-              })
-          )
-          .flatMap(IO.fromOption(_))
-          .mapError[ReadErrorsVector[String, String]](_ => singleton(ReadError.MissingValue(path)))
-          .provideLayer(System.live)
-      },
-      SystemEnvironment :: Nil
-    )
+  def fromSystemEnv(valueSeparator: Option[Char] = None): UIO[ConfigSource[String, String]] =
+    UIO
+      .effectTotal(sys.env)
+      .map(map => ConfigSource.fromMap(map, SystemEnvironment, '_', valueSeparator.getOrElse(':')))
 
-  def fromProperty(valueSeparator: Option[String] = None): ConfigSource[String, String] =
-    ConfigSource(
-      (path: Vector[String]) => {
-        val key = path.mkString(".")
-        system
-          .property(key)
-          .bimap(
-            ReadError.FatalError(path, _),
-            opt =>
-              opt.map(r => {
-                val consOfValues: ::[String] =
-                  valueSeparator
-                    .flatMap(
-                      sep =>
-                        r.split(sep).toList.map(_.trim) match {
-                          case h :: tail => Some(::(h, tail))
-                          case Nil       => None
-                        }
-                    )
-                    .getOrElse(::(r, Nil))
-                ConfigValue(consOfValues)
-              })
-          )
-          .flatMap(IO.fromOption(_))
-          .mapError[ReadErrorsVector[String, String]](_ => singleton(ReadError.MissingValue(path)))
-          .provideLayer(System.live)
-      },
-      SystemProperties :: Nil
+  def fromSystemProperties(valueSeparator: Option[Char] = None): UIO[ConfigSource[String, String]] =
+    for {
+      systemProperties <- UIO.effectTotal(java.lang.System.getProperties)
```

Comment:
We should create tickets in `zio` to expose all system properties and environment variables in maps, so we can just use those methods here in this project, without re-wrapping.

Reply:
But this is fine for now!